### PR TITLE
feat: allow math expressions for time and duration

### LINKED ISSUE #22009: [pkg/ottl] Mathematical Expressions are able to handle `time.Time` and `time.Duration` types.
In order to support complex interactions with time we need to support the following cases in our mathematical expressions:

- Subtracting a `time.Time` from `time.Time` must result in a `time.Duration`.
- Adding a `time.Duration` to a `time.Time` must result in a `time.Time`
- Subtracting a `time.Duration` from a `time.Time` must result in a `time.Time`
- Adding a `time.Duration` to a `time.Duration` must result in a `time.Duration`
- Subtracting a `time.Duration` from a `time.Duration` must result in a `time.Duration`

Only these type interactions should be supported. For example, the following SHOULD NOT be allowed:
- adding/subtracting an integer from a `time.Time`.
- adding/subtracting an integer from a `time.Duration`.

At the moment these rules cannot be enforced by the grammar, instead we will have to enforce these rules in the execution of the mathematical expressions.  This is the existing pattern we use for other type checking.

### PR DESCRIPTION
**Pull request by fchikwekwe**
Description: Allows addition and subtraction between times and duration according to rules in #22009

Link to tracking Issue: Closes #22009

Testing: Unit testing for math operations

Documentation:

### CODE REVIEW

File: pkg/ottl/math.go [hunk 1]
```diff
@@ -98,14 +99,58 @@ func attemptMathOperation[K any](lhs Getter[K], op mathOp, rhs Getter[K]) Getter
...
 				default:
-					return nil, fmt.Errorf("%v must be int64 or float64", x)
+					return nil, fmt.Errorf("%v must be int64, float64, time.Time or time.Duration", x)
 				}
 			},
 		},
 	}
 }
 
+func performOpTimeDuration(x any, y any, op mathOp) (any, error) {
```
[the remark this answers is not in the document]
Comment (fchikwekwe) [reply pkg/ottl/math.go]: Done! I've split the math helpers into two funcs for duration and time.

File: pkg/ottl/__debug_bin878371663
Comment (TylerHelmuth) [inline pkg/ottl/__debug_bin878371663:1]: This can be removed

File: pkg/ottl/README.md [hunk 2]
```diff
@@ -172,9 +172,17 @@ When defining an OTTL function, if the function needs to take an Enum then the f
 
 Math Expressions represent arithmetic calculations.  They support `+`, `-`, `*`, and `/`, along with `()` for grouping.
 
-Math Expressions currently only support `int64` and `float64`.
+Math Expressions currently support `int64` and`float64`. 
+Math Expressions also support `time.Time` and `time.Duration`, but only with `+` and `-` and only using the following rules:
```
Comment (TylerHelmuth) [inline pkg/ottl/README.md:176]:
```suggestion
Math Expressions currently support `int64`, `float64`, `time.Time` and `time.Duration`. 
For `time.Time` and `time.Duration`, only `+` and `-` are supported with the following rules: 
```

File: pkg/ottl/math.go [hunk 3]
```diff
@@ -98,14 +99,68 @@ func attemptMathOperation[K any](lhs Getter[K], op mathOp, rhs Getter[K]) Getter
...
+	case SUB:
+		switch newY := y.(type) {
+		case time.Time:
+			result := x.Sub(newY)
+			return result, nil
+		case time.Duration:
+			result := x.Add(-1 * newY)
+			return result, nil
+		default:
+			return nil, fmt.Errorf("time.Time must be subtracted from time.Time or time.Duration; found %v instead", y)
```
Comment (TylerHelmuth) [inline pkg/ottl/math.go:133]:
```suggestion
			return nil, fmt.Errorf("time.Time or time.Duration must be subtracted from time.Time; found %v instead", y)
```

**Review by TylerHelmuth — COMMENTED** [review]
Lets add some tests to `Test_evaluateMathExpression_error` that check that we get errors if we try to use incorrect operands or incorrect types.

**Comment by fchikwekwe** [comment]
@TylerHelmuth I've added some tests checking for various cases that should error. Let me know if you think I'm missing some cases.

**Review by TylerHelmuth — APPROVED** [review]

**Comment by TylerHelmuth** [comment]
@evan-bradley can you take a look at this one

File: pkg/ottl/README.md [hunk 4]
```diff
@@ -172,9 +172,17 @@ When defining an OTTL function, if the function needs to take an Enum then the f
 
 Math Expressions represent arithmetic calculations.  They support `+`, `-`, `*`, and `/`, along with `()` for grouping.
 
-Math Expressions currently only support `int64` and `float64`.
+Math Expressions currently support `int64`, `float64`, `time.Time` and `time.Duration`. 
+For `time.Time` and `time.Duration`, only `+` and `-` are supported with the following rules: 
+  - A `time.Time` `-` a `time.Time` yields a `time.Duration`.
+  - A `time.Duration` `+` a `time.Time` yields a `time.Time`. 
+  - A `time.Time` `-`  a `time.Duration` yields a `time.Time`.
```
Comment (evan-bradley) [inline pkg/ottl/README.md:179]:
```suggestion
  - A `time.Time` `+`  a `time.Duration` yields a `time.Time`.
  - A `time.Time` `-`  a `time.Duration` yields a `time.Time`.
```

Could we add `time + duration`? I get that the commutative case is already covered here, but ideally users don't have to think about the order.

Comment (fchikwekwe) [reply pkg/ottl/README.md:180]: Sounds great!

**Review by evan-bradley — APPROVED** [review]

**Comment by fchikwekwe** [comment]
@TylerHelmuth do you think we're ready for merge yet?

**Pull request merged**